programs/openbook-v2/src/state/orderbook/queue.rs

```
@@ -39,117 +46,190 @@ impl EventQueue {
     }
 }
 
 #[zero_copy]
-#[derive(bytemuck::Pod, bytemuck::Zeroable)]
 pub struct EventQueueHeader {
-    head: u32,
-    count: u32,
+    free_head: u16,
+    used_head: u16,
+    count: u16,
+    _padd: u16,
     pub seq_num: u64,
 }
 const_assert_eq!(std::mem::size_of::<EventQueueHeader>(), 16);
 const_assert_eq!(std::mem::size_of::<EventQueueHeader>() % 8, 0);
 
-impl QueueHeader for EventQueueHeader {
-    type Item = AnyEvent;
+impl EventQueueHeader {
+    pub fn count(&self) -> usize {
+        self.count as usize
+    }
+
+    pub fn free_head(&self) -> usize {
+        self.free_head as usize
+    }
 
-    fn head(&self) -> usize {
-        self.head as usize
+    pub fn used_head(&self) -> usize {
+        self.used_head as usize
     }
-    fn set_head(&mut self, value: u32) {
-        self.head = value;
+
+    fn set_free_head(&mut self, value: u16) {
+        self.free_head = value;
     }
-    fn count(&self) -> usize {
-        self.count as usize
+
+    fn set_used_head(&mut self, value: u16) {
+        self.used_head = value;
     }
-    fn set_count(&mut self, value: u32) {
-        self.count = value;
+
+    fn incr_count(&mut self) {
+        self.count += 1;
     }
+
+    fn decr_count(&mut self) {
+        self.count -= 1;
+    }
+
     fn incr_event_id(&mut self) {
         self.seq_num += 1;
     }
-    fn decr_event_id(&mut self, n: u64) {
-        self.seq_num -= n;
+}
+
+#[zero_copy]
+#[derive(Debug)]
+pub struct EventNode {
+    next: u16,
+    prev: u16,
+    _pad: [u8; 4],
+    pub event: AnyEvent,
+}
+const_assert_eq!(std::mem::size_of::<EventNode>(), 8 + 200);
+const_assert_eq!(std::mem::size_of::<EventNode>() % 8, 0);
+
+impl EventNode {
+    pub fn is_free(&self) -> bool {
+        self.prev == NULL
+    }
+
+    pub fn next(&self) -> usize {
+        self.next as usize
+    }
+
+    pub fn prev(&self) -> usize {
+        self.prev as usize
+    }
+
+    fn set_next(&mut self, next: usize) {
+        self.next = next as u16;
```

Currently there's a lot of `as u16` and `as usize`. Would always using `u16` for slot indexes work, including on these functions?

---

the problem is that to index arrays, rust requires `usize`, while the slots are stored as `u16` to save some space. I've tried to do less conversions but without success, so i ended up wrapping all the conversions in these methods

---

Yeah, painful. There could be inline accessor methods that have the u16 -> usize conversion? Otherwise I'd prefer if at least all narrowing conversions did `.try_into().unwrap()` so we get panics instead of silent truncation in case of bugs.

---

Great, will give it another look tomorrow!